src/charm.py

```diff
@@ -4,27 +4,263 @@
+                # bootstrap the data directory and users
+                logger.debug("Bootstrapping instance")
+                self._mysql.bootstrap_instance()
+
+                # Add the pebble layer
+                container.add_layer(self._mysqld_service, self._pebble_layer, combine=False)
+                container.restart(self._mysqld_service)
+                logger.debug("Waiting for instance to be ready")
+                self._mysql._wait_until_mysql_connection()
+
+                logger.info("Configuring instance")
+                # Configure all base users and revoke
+                # privileges from the root users
+                self._mysql.configure_mysql_users()
+                # Configure instance as a cluster node
+                self._mysql.configure_instance()
+
+            except (
+                MySQLConfigureInstanceError,
+                MySQLConfigureMySQLUsersError,
+                MySQLBootstrapInstanceError,
+                MySQLPatchDNSSearchesError,
+            ) as e:
+                self.unit.status = BlockedStatus("Unable to configure instance")
+                logger.debug("Unable to configure instance: {}".format(e))
+                return
+
+            if self.unit.is_leader():
+                try:
+                    # Create the cluster when
+                    self._mysql.create_cluster()
+                except MySQLCreateClusterError as e:
+                    self.unit.status = BlockedStatus("Unable to create cluster")
+                    logger.debug("Unable to create cluster: {}".format(e))
+            else:
+                # When unit is not the leader, it should wait
+                # for the leader to configure it a cluster node
+                self.unit.status = WaitingStatus("Waiting for peer relation join")
+
+            # Create control file in data directory
+            container.push(CONFIGURED_FILE, make_dirs=True, source="configured")
+
+        else:
+            # Configure the layer when changed
+            current_layer = container.get_plan()
+            new_layer = self._pebble_layer
+
+            if new_layer.services != current_layer:
+                logger.info("Add pebble layer")
+                container.add_layer(self._mysqld_service, new_layer, combine=True)
+                container.restart(self._mysqld_service)
+                self._mysql._wait_until_mysql_connection()
 
-    def _on_mysql_pebble_ready(self, _):
-        """Define and start a workload using the Pebble API."""
         self.unit.status = ActiveStatus()
 
+    def _on_peer_relation_joined(self, event: RelationJoinedEvent):
+        """Handle the peer relation joined event."""
+        # Only leader unit add instances to the cluster
+        if not self.unit.is_leader():
+            return
+
+        new_instance = self._get_hostname_by_unit(event.unit.name)
+
+        if not self._mysql.is_instance_configured_for_innodb(new_instance):
```

Comment:
I'm not sure if this is possible: the `peer_relation_joined` happens right after `peer_relation_created`, both during setup and before `leader_elected` event. Then calling `_mysql` would fail since it's not configured. A check for peer_data['configured'] would be needed.
As I said though, I don't know if the sequence of events can happen that way.

Reply:
In any case, this will be mitigate by this:
```python
if not self.unit.is_leader():
            return
```
above the instance adding.

Reply:
Ah, I missed that :)